fw/rbcx-coprocessor/src/CdcUartTunnel.cpp

```diff
@@ -51,51 +161,101 @@ void tunnelUartInit() {
     dmaTxHandle.Init.PeriphDataAlignment = DMA_PDATAALIGN_BYTE;
     dmaTxHandle.Init.Priority = DMA_PRIORITY_MEDIUM;
     HAL_DMA_Init(&dmaTxHandle);
+    HAL_NVIC_SetPriority(tunnelUartTxDmaIRQn, tunnelUartTxDmaIrqPrio, 0);
+    HAL_NVIC_EnableIRQ(tunnelUartTxDmaIRQn);
     LL_USART_EnableDMAReq_TX(tunnelUart);
-
-    pinInit(
-        tunnelUartTxPin, GPIO_MODE_AF_PP, GPIO_PULLUP, GPIO_SPEED_FREQ_HIGH);
-    pinInit(
-        tunnelUartRxPin, GPIO_MODE_AF_INPUT, GPIO_PULLUP, GPIO_SPEED_FREQ_HIGH);
-}
-
-static void primaryUartRxPoll() {
-    int rxHead = rxFifo.size() - __HAL_DMA_GET_COUNTER(&dmaRxHandle);
-    rxFifo.setHead(rxHead);
 }
 
 static void primaryUartTx(uint8_t* data, size_t len) {
-    HAL_DMA_Start(&dmaTxHandle, uint32_t(data), uint32_t(&tunnelUart->DR), len);
+    dmaTxComplete = false;
+    HAL_DMA_Start_IT(
+        &dmaTxHandle, uint32_t(data), uint32_t(&tunnelUart->DR), len);
 }
 
-static bool primaryUartTxReady() {
-    HAL_DMA_PollForTransfer(&dmaTxHandle, HAL_DMA_FULL_TRANSFER, 0);
-    return dmaTxHandle.State == HAL_DMA_STATE_READY;
+static void tunnelDownstreamHandler() {
+    if (!dmaTxComplete)
+        return;
```

Comment:
Tohle funguje jinak, než logika s *HAL_DMA_PollForTransfer* ? Nešlo by to vyřešit prostou logikou nad *State* nebo *Status* ?

Reply:
Takže.

`HAL_DMA_PollForTransfer` není "poll for transfer", je to to "finish transfer". Sleduj, co se stane, když DMA není hotové, ale hitneš timeout (i ten náš 0): https://github.com/STMicroelectronics/STM32CubeF1/blob/441b2cbdc25aa50437a59c4bffe22b88e78942c9/Drivers/STM32F1xx_HAL_Driver/Src/stm32f1xx_hal_dma.c#L566

Tu metodu, `HAL_DMA_PollForTransfer`, nejde použít na polling.

Fix je ten řádek zakomentovat. Udělal jsem v HalDma.hpp metodu `HAL_DMA_PollForTransfer_Really`, která to tak má, s náležitým komentářem.
 https://github.com/RoboticsBrno/RB3204-RBCX/pull/31/files#diff-3a801033478339a5be5fb1d08092ff27

Reply:
Bych prostě čekoval `if (status == HAL_OK && dmaTxHandle.State == READY`

Reply:
To ale nejde, protože ten timeout nastaví `State` == READY, a další volání `HAL_DMA_PollForTransfer` už failne tady a OK nedostaneš nikdy https://github.com/STMicroelectronics/STM32CubeF1/blob/441b2cbdc25aa50437a59c4bffe22b88e78942c9/Drivers/STM32F1xx_HAL_Driver/Src/stm32f1xx_hal_dma.c#L507

Reply:
Achjo